File: src/companies_house_abm/abm/agents/bank.py
"""Bank agent for the ABM.

Banks accept deposits, extend credit to firms, and must satisfy
regulatory capital and reserve requirements.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from companies_house_abm.abm.agents.base import BaseAgent

if TYPE_CHECKING:
    from typing import Any

    import numpy as np

    from companies_house_abm.abm.config import BankBehaviorConfig, BankConfig


class Bank(BaseAgent):
    """A bank agent.

    Attributes:
        capital: Bank equity / own funds.
        reserves: Central bank reserves held.
        loans: Total outstanding loans.
        deposits: Total deposits held.
        non_performing_loans: Loans in default.
        interest_rate: Rate charged on new loans.
        profit: Net income for the period.
    """

    def __init__(
        self,
        agent_id: str | None = None,
        *,
        capital: float = 0.0,
        reserves: float = 0.0,
        loans: float = 0.0,
        deposits: float = 0.0,
        config: BankConfig | None = None,
        behavior: BankBehaviorConfig | None = None,
    ) -> None:
        super().__init__(agent_id)
        self.capital = capital
        self.reserves = reserves
        self.loans = loans
        self.deposits = deposits
        self.non_performing_loans: float = 0.0
        self.interest_rate: float = 0.05
        self.profit: float = 0.0
        self._interest_income: float = 0.0
        self._interest_expense: float = 0.0

        self._config = config
        self._behavior = behavior

    # ------------------------------------------------------------------
    # Regulatory ratios
    # ------------------------------------------------------------------

    @property
    def capital_ratio(self) -> float:
        """Capital adequacy ratio (capital / risk-weighted assets)."""
        rw = self._config.risk_weight if self._config else 1.0
        risk_weighted = self.loans * rw
        if risk_weighted <= 0:
            return 1.0
        return self.capital / risk_weighted

    @property
    def reserve_ratio(self) -> float:
        """Reserve ratio (reserves / deposits)."""
        if self.deposits <= 0:
            return 1.0
        return self.reserves / self.deposits

    @property
    def meets_capital_requirement(self) -> bool:
        """Whether the bank satisfies the regulatory capital requirement."""
        req = self._config.capital_requirement if self._config else 0.10
        buffer = self._behavior.capital_buffer if self._behavior else 0.02
        return self.capital_ratio >= req + buffer

    # ------------------------------------------------------------------
    # Step logic
    # ------------------------------------------------------------------

    def step(self) -> None:
        """Execute one period of bank behaviour.

        1. Set lending rate based on policy rate and risk.
        2. Calculate interest income/expense.
        3. Update profit and capital.
        """
        self._set_lending_rate(policy_rate=0.05)
        self._calculate_income()
        self._update_capital()

    def set_policy_rate(self, rate: float) -> None:
        """Update the lending rate using a new policy rate.

        Args:
            rate: The central bank policy rate.
        """
        self._set_lending_rate(rate)

    def _set_lending_rate(self, policy_rate: float) -> None:
        """Determine the interest rate for new loans."""
        markup = self._behavior.base_interest_markup if self._behavior else 0.02
        risk = self._behavior.risk_premium_sensitivity if self._behavior else 0.05
        npl_ratio = self.non_performing_loans / self.loans if self.loans > 0 else 0.0
        self.interest_rate = policy_rate + markup + risk * npl_ratio

    def _calculate_income(self) -> None:
        """Calculate interest income and expense for the period."""
        self._interest_income = self.interest_rate * self.loans
        deposit_rate = max(self.interest_rate - 0.02, 0.0)
        self._interest_expense = deposit_rate * self.deposits

    def _update_capital(self) -> None:
        """Update profit and capital after the period."""
        provisions = self.non_performing_loans * 0.5
        self.profit = self._interest_income - self._interest_expense - provisions
        self.capital += self.profit

    # ------------------------------------------------------------------
    # Lending interface used by the credit market
    # ------------------------------------------------------------------

    def evaluate_loan(
        self,
        amount: float,
        borrower_equity: float,
        borrower_revenue: float,
        rng: np.random.Generator | None = None,
    ) -> bool:
        """Decide whether to extend a loan.

        When ``credit_score_noise_std > 0`` and an RNG is provided, the
        decision uses a bounded-rationality composite credit score (Gabaix
        2014): each criterion is normalised to its threshold value and the
        weighted average is perturbed by Gaussian noise before comparing
        against a score of 1.0.  This models lender imprecision and produces
        realistic cross-sectional dispersion in credit outcomes.

        When ``credit_score_noise_std == 0`` or no RNG is given, the original
        deterministic hard-threshold decision rule is used.

        Args:
            amount: Requested loan amount.
            borrower_equity: Borrower's equity / net assets.
            borrower_revenue: Borrower's revenue.
            rng: Optional random number generator for noisy scoring.

        Returns:
            True if the loan is approved.
        """
        if not self.meets_capital_requirement:
            return False

        collateral_req = 0.5
        threshold = self._behavior.lending_threshold if self._behavior else 0.3
        noise_std = self._behavior.credit_score_noise_std if self._behavior else 0.0

        if borrower_revenue <= 0:
            return False

        if noise_std > 0 and rng is not None:
            # Bounded rationality: composite score with Gaussian noise
            # Each component is normalised so that 1.0 = exactly at threshold.
            collateral_score = borrower_equity / max(amount * collateral_req, 1e-9)
            coverage_score = (
                borrower_revenue / max(amount * self.interest_rate, 1e-9)
            ) / max(threshold, 1e-9)
            composite = 0.5 * collateral_score + 0.5 * coverage_score
            noise = float(rng.normal(0, noise_std))
            return composite + noise > 1.0

        # Deterministic hard-threshold rule (original behaviour)
        if borrower_equity < amount * collateral_req:
            return False
        debt_service_coverage = borrower_revenue / max(
            amount * self.interest_rate, 1e-9
        )
        return debt_service_coverage >= threshold

    def extend_loan(self, amount: float) -> float:
        """Extend a loan and return the interest rate charged.

        Args:
            amount: The loan principal.

        Returns:
            The interest rate on the loan.
        """
        self.loans += amount
        self.deposits += amount  # loan creates a deposit
        return self.interest_rate

    def record_default(self, amount: float) -> None:
        """Record a loan default.

        Args:
            amount: The defaulted loan amount.
        """
        self.non_performing_loans += amount

    def record_repayment(self, amount: float) -> None:
        """Record a loan repayment.

        Args:
            amount: The repayment amount.
        """
        self.loans = max(self.loans - amount, 0.0)

    # ------------------------------------------------------------------
    # State reporting
    # ------------------------------------------------------------------

    def get_state(self) -> dict[str, Any]:
        """Return a snapshot of the bank's state."""
        return {
            "agent_id": self.agent_id,
            "agent_type": self.agent_type,
            "capital": self.capital,
            "reserves": self.reserves,
            "loans": self.loans,
            "deposits": self.deposits,
            "non_performing_loans": self.non_performing_loans,
            "interest_rate": self.interest_rate,
            "capital_ratio": self.capital_ratio,
            "profit": self.profit,
        }

File: packages/rust-abm/src/agents/bank.rs
use std::fmt;

use krabmaga::engine::{agent::Agent, state::State};

use crate::state::EconomyState;

// ─────────────────────────────────────────────────────────────────────────────
// Bank data
// ─────────────────────────────────────────────────────────────────────────────

/// All mutable state for a single bank.
#[derive(Clone, Debug)]
pub struct BankData {
    pub capital: f64,
    pub reserves: f64,
    pub loans: f64,
    pub deposits: f64,
    pub non_performing_loans: f64,
    pub interest_rate: f64,
    pub profit: f64,
    // Internal per-period income calculations
    pub interest_income: f64,
    pub interest_expense: f64,
}

impl BankData {
    pub fn new(capital: f64, reserves: f64) -> Self {
        BankData {
            capital,
            reserves,
            loans: 0.0,
            deposits: 0.0,
            non_performing_loans: 0.0,
            interest_rate: 0.05,
            profit: 0.0,
            interest_income: 0.0,
            interest_expense: 0.0,
        }
    }

    // ─── Regulatory ratios ──────────────────────────────────────────────────

    pub fn capital_ratio(&self, risk_weight: f64) -> f64 {
        let risk_weighted = self.loans * risk_weight;
        if risk_weighted <= 0.0 {
            1.0
        } else {
            self.capital / risk_weighted
        }
    }

    pub fn meets_capital_requirement(
        &self,
        capital_requirement: f64,
        capital_buffer: f64,
        risk_weight: f64,
    ) -> bool {
        self.capital_ratio(risk_weight) >= capital_requirement + capital_buffer
    }

    // ─── Step sub-methods ───────────────────────────────────────────────────

    pub fn set_lending_rate(&mut self, policy_rate: f64, base_markup: f64, risk: f64) {
        let npl_ratio = if self.loans > 0.0 {
            self.non_performing_loans / self.loans
        } else {
            0.0
        };
        self.interest_rate = policy_rate + base_markup + risk * npl_ratio;
    }

    pub fn calculate_income(&mut self) {
        self.interest_income = self.interest_rate * self.loans;
        let deposit_rate = (self.interest_rate - 0.02).max(0.0);
        self.interest_expense = deposit_rate * self.deposits;
    }

    pub fn update_capital(&mut self) {
        let provisions = self.non_performing_loans * 0.5;
        self.profit = self.interest_income - self.interest_expense - provisions;
        self.capital += self.profit;
    }

    pub fn step(&mut self, policy_rate: f64, base_markup: f64, risk: f64) {
        self.set_lending_rate(policy_rate, base_markup, risk);
        self.calculate_income();
        self.update_capital();
    }

    // ─── Lending interface ──────────────────────────────────────────────────

    pub fn evaluate_loan(
        &self,
        amount: f64,
        borrower_equity: f64,
        borrower_revenue: f64,
        capital_requirement: f64,
        capital_buffer: f64,
        risk_weight: f64,
        lending_threshold: f64,
    ) -> bool {
        if !self.meets_capital_requirement(capital_requirement, capital_buffer, risk_weight) {
            return false;
        }
        let collateral_req = 0.5;
        if borrower_equity < amount * collateral_req {
            return false;
        }
        if borrower_revenue <= 0.0 {
            return false;
        }
        let debt_service = borrower_revenue / (amount * self.interest_rate).max(1e-9);
        debt_service >= lending_threshold
    }

    pub fn extend_loan(&mut self, amount: f64) -> f64 {
        self.loans += amount;
        self.deposits += amount; // loan creates deposit
        self.interest_rate
    }

    pub fn record_default(&mut self, amount: f64) {
        self.non_performing_loans += amount;
    }

    pub fn record_repayment(&mut self, amount: f64) {
        self.loans = (self.loans - amount).max(0.0);
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Central bank data
// ─────────────────────────────────────────────────────────────────────────────

/// All mutable state for the central bank (singleton).
#[derive(Clone, Debug)]
pub struct CentralBankData {
    pub policy_rate: f64,
    pub inflation_target: f64,
    pub current_inflation: f64,
    pub output_gap: f64,
    pub previous_rate: f64,
}

impl CentralBankData {
    pub fn new(inflation_target: f64) -> Self {
        CentralBankData {
            policy_rate: inflation_target,
            inflation_target,
            current_inflation: inflation_target,
            output_gap: 0.0,
            previous_rate: inflation_target,
        }
    }

    /// Apply Taylor rule and update the policy rate.
    pub fn step(
        &mut self,
        pi_coef: f64,
        y_coef: f64,
        smoothing: f64,
        lower: f64,
    ) {
        let target_rate = self.inflation_target
            + pi_coef * (self.current_inflation - self.inflation_target)
            + y_coef * self.output_gap;

        let smoothed = smoothing * self.previous_rate + (1.0 - smoothing) * target_rate;
        self.previous_rate = self.policy_rate;
        self.policy_rate = smoothed.max(lower);
    }

    pub fn update_observations(&mut self, inflation: f64, output_gap: f64) {
        self.current_inflation = inflation;
        self.output_gap = output_gap;
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Government data
// ─────────────────────────────────────────────────────────────────────────────

/// All mutable state for the government (singleton).
#[derive(Clone, Debug)]
pub struct GovernmentData {
    pub tax_revenue: f64,
    pub expenditure: f64,
    pub transfer_spending: f64,
    pub deficit: f64,
    pub debt: f64,
    pub gdp_estimate: f64,
}

impl GovernmentData {
    pub fn new() -> Self {
        GovernmentData {
            tax_revenue: 0.0,
            expenditure: 0.0,
            transfer_spending: 0.0,
            deficit: 0.0,
            debt: 0.0,
            gdp_estimate: 0.0,
        }
    }

    pub fn begin_period(&mut self) {
        self.tax_revenue = 0.0;
        self.expenditure = 0.0;
        self.transfer_spending = 0.0;
    }

    pub fn collect_corporate_tax(&mut self, profits: f64, rate: f64) -> f64 {
        let tax = (profits * rate).max(0.0);
        self.tax_revenue += tax;
        tax
    }

    pub fn collect_income_tax(&mut self, income: f64, rate: f64) -> f64 {
        let tax = (income * rate).max(0.0);
        self.tax_revenue += tax;
        tax
    }

    pub fn calculate_spending(&mut self, spending_gdp_ratio: f64) -> f64 {
        self.expenditure = spending_gdp_ratio * self.gdp_estimate.max(0.0);
        self.expenditure
    }

    pub fn pay_unemployment_benefit(
        &mut self,
        average_wage: f64,
        unemployed_count: usize,
        replacement: f64,
    ) -> f64 {
        let total = replacement * average_wage * unemployed_count as f64;
        self.transfer_spending += total;
        total
    }

    pub fn apply_fiscal_rule(&mut self, deficit_target: f64, speed: f64) {
        if self.gdp_estimate <= 0.0 {
            return;
        }
        let current_deficit_ratio =
            self.deficit.abs() / self.gdp_estimate.max(1e-9);
        let gap = current_deficit_ratio - deficit_target;
        let adjustment = speed * gap * self.gdp_estimate;
        self.expenditure = (self.expenditure - adjustment).max(0.0);
    }

    pub fn end_period(&mut self) {
        self.deficit = self.tax_revenue - (self.expenditure + self.transfer_spending);
        self.debt -= self.deficit; // negative deficit → debt increases
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// krabmaga Agent proxy for Bank
// ─────────────────────────────────────────────────────────────────────────────

/// Proxy agent for a single bank.
#[derive(Clone)]
pub struct BankAgent {
    pub id: usize,
}

impl fmt::Display for BankAgent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "BankAgent({})", self.id)
    }
}

impl Agent for BankAgent {
    /// Execute one period of bank behaviour.
    fn step(&mut self, state: &mut dyn State) {
        let state = state
            .as_any_mut()
            .downcast_mut::<EconomyState>()
            .expect("state should be EconomyState");

        let policy_rate = state.central_bank.policy_rate;
        let markup = state.config.base_interest_markup;
        let risk = state.config.risk_premium_sensitivity;

        let bank = &mut state.banks[self.id];
        bank.step(policy_rate, markup, risk);
    }
}
